### PR TITLE
Overlink I/O: Read and Write VAR_ATTRIBUTES

### PR DESCRIPTION
**Pull request by JustinPrivitera**
Addresses a bullet in #1111

Also adds volume dependence to blueprint index, if available.

### CODE REVIEW

File: src/libs/relay/conduit_relay_io_silo.cpp [hunk 1]
```diff
@@ -1944,6 +1989,84 @@ read_state(DBfile *dbfile,
...
+{
+    if (! root_node[multimesh_name].has_child("vars"))
+    {
+        // nothing to do if there are no vars
+        return;
+    }
+
+    // TODO it is annoying to see the error message that this was not found in the
+    // silo file every time. Is there a way to check that it is there OR to not call this
+    // unconditionally?
```
Comment (cyrush) [inline src/libs/relay/conduit_relay_io_silo.cpp]: I think we can use `DBInqVarExists` and `DBInqVarType` to check if this is there and then make sure it is the type we expect.. 

I think that would remove the error messages.

File: src/libs/relay/conduit_relay_io_silo.cpp [hunk 2]
```diff
@@ -4003,6 +4154,103 @@ write_multimats(DBfile *dbfile,
...
+
+            elemlengths.push_back(num_attr);
+            nvalues += num_attr;
+
+            // 
+            // centering: ATTR_NODAL 0, ATTR_ZONAL 1, ATTR_FACE, ATTR_EDGE
+            // 
+            if (n_var["association"].as_string() == "vertex")
+            {
+                var_attr_values.push_back(0); // nodal == vertex
```
Comment (cyrush) [inline src/libs/relay/conduit_relay_io_silo.cpp:4202]: Would it help to use the Silo enum types here, instead of the raw values?

**Review by cyrush — APPROVED** [review]
Looks great!

Two minor suggestions, one addressing how to avoid the error messages when var atts doesn't exist.

File: src/libs/relay/conduit_relay_io_silo.cpp (hunk 1 above)
Comment (JustinPrivitera) [reply src/libs/relay/conduit_relay_io_silo.cpp]: I will add that here, and maybe in a follow up PR add that elsewhere when reading other variables that may not be present.

File: src/libs/relay/conduit_relay_io_silo.cpp (hunk 2 above)
Comment (JustinPrivitera) [reply src/libs/relay/conduit_relay_io_silo.cpp:4202]: We don't have access to the raw enum types, as these are Overlink enum values. Even if Overlink and Silo agree on the enum values and meanings (not sure if they do for centering or not), I think it is cleaner to use raw values and clearly label what the Overlink values mean.

**Comment by cyrush** [comment]
My bad , I forgot they were not  proper silo enums. Sorry!

-Cyrus
________________________________

**Pull request merged**